File: app/src/main/java/com/krystianwsul/checkme/gui/instances/ShowNotificationGroupActivity.java
package com.krystianwsul.checkme.gui.instances;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.support.v7.app.ActionBar;
import android.support.v7.view.ActionMode;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import com.krystianwsul.checkme.R;
import com.krystianwsul.checkme.gui.AbstractActivity;
import com.krystianwsul.checkme.loaders.ShowNotificationGroupLoader;
import com.krystianwsul.checkme.utils.InstanceKey;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Set;

public class ShowNotificationGroupActivity extends AbstractActivity implements GroupListFragment.GroupListListener, LoaderManager.LoaderCallbacks<ShowNotificationGroupLoader.Data> {
    private static final String INSTANCES_KEY = "instanceKeys";

    private Set<InstanceKey> mInstanceKeys;

    private GroupListFragment mGroupListFragment;

    private boolean mSelectAllVisible = false;

    @NonNull
    public static Intent getIntent(@NonNull Context context, @NonNull ArrayList<InstanceKey> instanceKeys) {
        Assert.assertTrue(!instanceKeys.isEmpty());

        Intent intent = new Intent(context, ShowNotificationGroupActivity.class);
        //intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        intent.putParcelableArrayListExtra(INSTANCES_KEY, instanceKeys);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_show_notification_group);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        Assert.assertTrue(toolbar != null);

        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        Assert.assertTrue(actionBar != null);

        actionBar.setTitle(null);

        mGroupListFragment = (GroupListFragment) getSupportFragmentManager().findFragmentById(R.id.show_notification_group_list);
        if (mGroupListFragment == null) {
            mGroupListFragment = GroupListFragment.newInstance();

            getSupportFragmentManager().beginTransaction()
                    .add(R.id.show_notification_group_list, mGroupListFragment)
                    .commit();
        }

        Intent intent = getIntent();
        Assert.assertTrue(intent.hasExtra(INSTANCES_KEY));
        ArrayList<InstanceKey> instanceKeys = intent.getParcelableArrayListExtra(INSTANCES_KEY);
        Assert.assertTrue(instanceKeys != null);
        Assert.assertTrue(!instanceKeys.isEmpty());

        mInstanceKeys = new HashSet<>(instanceKeys);

        getSupportLoaderManager().initLoader(0, null, this);
    }

    @Override
    public Loader<ShowNotificationGroupLoader.Data> onCreateLoader(int id, Bundle args) {
        return new ShowNotificationGroupLoader(this, mInstanceKeys);
    }

    @Override
    public void onLoadFinished(Loader<ShowNotificationGroupLoader.Data> loader, ShowNotificationGroupLoader.Data data) {
        Assert.assertTrue(data != null);

        mGroupListFragment.setInstanceKeys(mInstanceKeys, data.DataId, data.mDataWrapper);
    }

    @Override
    public void onLoaderReset(Loader<ShowNotificationGroupLoader.Data> loader) {

    }

    @Override
    public void onCreateGroupActionMode(ActionMode actionMode) {

    }

    @Override
    public void onDestroyGroupActionMode() {

    }

    @Override
    public void setGroupSelectAllVisibility(Integer position, boolean selectAllVisible) {
        mSelectAllVisible = selectAllVisible;

        invalidateOptionsMenu();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_select_all, menu);
        return true;
    }

    @Override
    public boolean onPrepareOptionsMenu(Menu menu) {
        menu.findItem(R.id.action_select_all).setVisible(mSelectAllVisible);

        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        Assert.assertTrue(item.getItemId() == R.id.action_select_all);
        Assert.assertTrue(mGroupListFragment != null);

        mGroupListFragment.selectAll();

        return true;
    }
}

File: app/src/main/java/com/krystianwsul/checkme/gui/instances/ShowNotificationGroupActivity.kt
package com.krystianwsul.checkme.gui.instances

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.os.Bundle
import androidx.appcompat.view.ActionMode
import com.krystianwsul.checkme.R
import com.krystianwsul.checkme.databinding.ActivityShowNotificationGroupBinding
import com.krystianwsul.checkme.databinding.BottomBinding
import com.krystianwsul.checkme.domainmodel.extensions.clearTaskEndTimeStamps
import com.krystianwsul.checkme.domainmodel.extensions.setTaskEndTimeStamps
import com.krystianwsul.checkme.domainmodel.update.AndroidDomainUpdater
import com.krystianwsul.checkme.gui.base.AbstractActivity
import com.krystianwsul.checkme.gui.dialogs.RemoveInstancesDialogFragment
import com.krystianwsul.checkme.gui.instances.list.GroupListListener
import com.krystianwsul.checkme.gui.instances.list.GroupListParameters
import com.krystianwsul.checkme.gui.instances.list.GroupListViewModel
import com.krystianwsul.checkme.gui.tree.AbstractHolder
import com.krystianwsul.checkme.gui.utils.BottomFabMenuDelegate
import com.krystianwsul.checkme.gui.utils.CopyAllRemindersDelegate
import com.krystianwsul.checkme.utils.startDate
import com.krystianwsul.checkme.utils.tryGetFragment
import com.krystianwsul.checkme.viewmodels.DataId
import com.krystianwsul.checkme.viewmodels.ShowNotificationGroupViewModel
import com.krystianwsul.checkme.viewmodels.getViewModel
import com.krystianwsul.common.utils.TaskKey
import com.krystianwsul.treeadapter.TreeViewAdapter
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.kotlin.addTo
import io.reactivex.rxjava3.kotlin.plusAssign
import java.io.Serializable

class ShowNotificationGroupActivity : AbstractActivity(), GroupListListener {

    companion object {

        private const val TAG_DELETE_INSTANCES = "deleteInstances"

        private const val KEY_MENU_DELEGATE_STATE = "menuDelegateState"

        fun getIntent(context: Context) = Intent(context, ShowNotificationGroupActivity::class.java)
    }

    private var selectAllVisible = false

    private val showNotificationGroupViewModel by lazy { getViewModel<ShowNotificationGroupViewModel>() }

    override val snackbarParent get() = binding.showNotificationGroupCoordinator

    private val deleteInstancesListener: (Serializable, Boolean) -> Unit = { taskKeys, removeInstances ->
        @Suppress("UNCHECKED_CAST")
        AndroidDomainUpdater.setTaskEndTimeStamps(
            showNotificationGroupViewModel.dataId.toFirst(),
            taskKeys as Set<TaskKey>,
            removeInstances,
        )
            .observeOn(AndroidSchedulers.mainThread())
            .flatMapMaybe { showSnackbarRemovedMaybe(it.taskKeys.size).map { _ -> it } }
            .flatMapCompletable {
                AndroidDomainUpdater.clearTaskEndTimeStamps(showNotificationGroupViewModel.dataId.toFirst(), it)
            }
            .subscribe()
            .addTo(createDisposable)
    }

    private val receiver = object : BroadcastReceiver() {

        override fun onReceive(context: Context?, intent: Intent?) = showNotificationGroupViewModel.refresh()
    }

    private val filterCriteria by lazy {
        binding.showNotificationGroupToolbarCollapseInclude
            .collapseAppBarLayout
            .searchParamsObservable
            .map { it.toFilterCriteria() } // todo connect
    }

    override val instanceSearch by lazy {
        filterCriteria.map { it.toExpandOnly() }
    }

    private var data: ShowNotificationGroupViewModel.Data? = null

    private lateinit var binding: ActivityShowNotificationGroupBinding
    private lateinit var bottomBinding: BottomBinding

    private lateinit var bottomFabMenuDelegate: BottomFabMenuDelegate

    override val groupListViewModel by lazy { getViewModel<GroupListViewModel>() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityShowNotificationGroupBinding.inflate(layoutInflater)
        bottomBinding = BottomBinding.bind(binding.root)
        setContentView(binding.root)

        bottomFabMenuDelegate = BottomFabMenuDelegate(
            bottomBinding,
            binding.showNotificationGroupCoordinator,
            this,
            savedInstanceState?.getParcelable(KEY_MENU_DELEGATE_STATE),
        )

        binding.groupListFragment.listener = this

        binding.showNotificationGroupToolbarCollapseInclude
            .collapseAppBarLayout
            .apply {
                setSearchMenuOptions(false, true, false)

                configureMenu(R.menu.show_task_menu_top, R.id.actionShowTaskSearch)

                menu.findItem(R.id.actionTaskShowDeleted).isVisible = false
            }

        updateTopMenu()
        initBottomBar()

        showNotificationGroupViewModel.apply {
            start()

            createDisposable += filterCriteria.map { it.search }.subscribe(searchRelay)

            createDisposable += data.subscribe {
                this@ShowNotificationGroupActivity.data = it

                val immediate = it.immediate

                binding.groupListFragment.setParameters(
                    GroupListParameters.InstanceKeys(
                        showNotificationGroupViewModel.dataId,
                        immediate,
                        it.groupListDataWrapper,
                    )
                )

                updateTopMenu()
            }
        }

        tryGetFragment<RemoveInstancesDialogFragment>(TAG_DELETE_INSTANCES)?.listener = deleteInstancesListener

        startDate(receiver)

        CopyAllRemindersDelegate(this, groupListViewModel, createDisposable)
    }

    private fun updateTopMenu() {
        binding.showNotificationGroupToolbarCollapseInclude
            .collapseAppBarLayout
            .menu
            .findItem(R.id.actionShowTaskSearch)
            .isVisible = !data?.groupListDataWrapper
            ?.allInstanceDatas
            .isNullOrEmpty()
    }

    override fun onStart() {
        super.onStart()

        binding.groupListFragment.checkCreatedTaskKey()
    }

    override fun onDestroy() {
        unregisterReceiver(receiver)

        super.onDestroy()
    }

    override fun onCreateGroupActionMode(
        actionMode: ActionMode,
        treeViewAdapter: TreeViewAdapter<AbstractHolder>,
        initial: Boolean,
    ) = Unit

    override fun onDestroyGroupActionMode() = Unit

    override fun setGroupMenuItemVisibility(position: Int?, selectAllVisible: Boolean) {
        this.selectAllVisible = selectAllVisible

        updateBottomBar()
    }

    private fun updateBottomBar() {
        bottomBinding.bottomAppBar
            .menu
            .findItem(R.id.action_select_all)
            ?.isVisible = selectAllVisible
    }

    override fun getBottomBar() = bottomBinding.bottomAppBar

    override fun initBottomBar() {
        bottomBinding.bottomAppBar.apply {
            replaceMenu(R.menu.menu_select_all)

            setOnMenuItemClickListener { item ->
                when (item.itemId) {
                    R.id.action_select_all -> binding.groupListFragment
                        .treeViewAdapter
                        .selectAll()
                    else -> throw IllegalArgumentException()
                }

                true
            }
        }

        updateBottomBar()
    }

    override fun deleteTasks(dataId: DataId, taskKeys: Set<TaskKey>) {
        RemoveInstancesDialogFragment.newInstance(taskKeys)
            .also { it.listener = deleteInstancesListener }
            .show(supportFragmentManager, TAG_DELETE_INSTANCES)
    }

    override fun showFabMenu(menuDelegate: BottomFabMenuDelegate.MenuDelegate) = bottomFabMenuDelegate.showMenu(menuDelegate)

    override fun setToolbarExpanded(expanded: Boolean) = binding.showNotificationGroupToolbarCollapseInclude
        .collapseAppBarLayout
        .setExpanded(expanded)

    override fun onBackPressed() {
        binding.showNotificationGroupToolbarCollapseInclude
            .collapseAppBarLayout
            .apply { if (isSearching) closeSearch() else super.onBackPressed() }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)

        outState.putParcelable(KEY_MENU_DELEGATE_STATE, bottomFabMenuDelegate.state)
    }
}